### DISCUSSION TITLE
[Adonis 4.1] App console.log still printing after break app execution

### DISCUSSION THREAD

**Discussion by wemersonrv** [adonisjs/core#1410] [Help] [answered]
Hello,

After break an `adonis serve --dev` command, and run other commands, like `adonis migration:run` the app still running, because the other commands not release the prompt for me, and the app `console.log` still printing log entries...

How can i check if the app still running after break the `adonis serve` command?

**Comment by McSneaky** [comment]
Where are log entries coming from? `adonis serve` command or `adonis migration` command?
If you want to programmatically kill it off you can use something like[ `process.exit()`](https://nodejs.org/api/process.html#process_process_exit_code)

Or if you killed it (CTRL + C), which released terminal, but serveri is still running you need to use some kind of task manager depending on your OS to find running NodeJS instance and kill that one

If CTRL + C has no effect at all, then most likely there is something eating off kill signal (bash wrapper or something?) and kill signal never reaches NodeJS process

Perhaps you just moved `adonis serve --dev` output from screen to background or started process in the background with printing output to screen also?

**Reply by wemersonrv** [reply]
> Where are log entries coming from? `adonis serve` command or `adonis migration` command?
> If you want to programmatically kill it off you can use something like[ `process.exit()`](https://nodejs.org/api/process.html#process_process_exit_code)
> 
> Or if you killed it (CTRL + C), which released terminal, but serveri is still running you need to use some kind of task manager depending on your OS to find running NodeJS instance and kill that one
> 
> If CTRL + C has no effect at all, then most likely there is something eating off kill signal (bash wrapper or something?) and kill signal never reaches NodeJS process
> 
> Perhaps you just moved `adonis serve --dev` output from screen to background or started process in the background with printing output to screen also?

On screen, the `console.log` prnts on screen... so when i stop the app (`control + C`) and confirm the stop... when run other commands, the console.log app continues to print... and not release the OS prompt for me... as it apps continues to works.

See:

![image](https://user-images.githubusercontent.com/2028673/89665000-5cbcaf00-d8ae-11ea-9c1b-0aa7fc4669c2.png)

1. App running
2. App stopped (control + c) and OS prompt released, and run other cli command. See the `adonis migration:refresh` logs printed on screen
3. After cli command ends, the app logs continues.

I don't have any screen with app running... no one terminal window... nothing!

# Edited: Found the problem
Thesese app log are from event listeners of a lib added to `IoC` as a singleton, and started at `boot`  on service provider. So the listeners appears to continue working after app breaked with `Control + C`.

**Comment by wemersonrv** [comment, marked as answer]
Make new tests here, and every adonis cli command i run, even if i run only `adonis` to show all cli options, it calls the `register` and `boot` from all servide providers... 

Well, i need to add these providers to app.. this is a fact. But only when the app is running, not when i just run a cli command. 

So, there is a way to check check for specific CLI commands i want to accept servicce providers `register` and `boot` executions and others no?

Example: If i run a new migration, service providers ran the `boot` method and i don't want that. But, if i ran a custom command i created to make some admin scheduled tasks in background, i need the services registered (and their singleton instances created at boot).

For now, there is no problem with this behavior because my project is not production ready yet. But, when its goes to production, i don't want to deal with these issues every new deploy cycle, Imagine if i have 100... 200 container instances of this project! Pure headache!

**Reply by thetutlage** [reply]
Providers are meant to registered and booted every time the app is loaded, regardless whether it is an HTTP server or not. For example: Migration command alone needs Lucid to be ready and Lucid needs other stuff like Config, Env and so on.

I am not sure, what exactly you are doing inside providers that is causing this issue. In v4, we have `isAceCommand` helper https://adonisjs.com/docs/4.0/helpers#_isacecommand to know, if the app is booted as an ace command or the HTTP server. You can use this flag to decide, if your code inside the `boot` method should be running or not.

Also, it is always nice to share concrete examples and code. Coz answers and work-arounds heavily rely on the real world examples.

**Reply by wemersonrv** [reply]
Hi @thetutlage, thanks for your help. `Helpers.isAceCommand()`  works.

Just add it to my servide providers... See the project repo: https://github.com/wemersonrv/whatsapp-teste

The project is starting yet, i'm just make some tests for now and it does not have much conen.

You can see on this line i'm using isAceCommand and is works for now. See it below:

https://github.com/wemersonrv/whatsapp-teste/blob/8ae9f68572b5923dbb0a6f0730490e6ebf02a863/providers/WhatsappProvider.js#L45

**Reply by thetutlage** [reply]
Since, you have mentioned that you have just started working on the project. I recommend using v5 over v4

**Reply by wemersonrv** [reply]
Adonis V5 is a good idea!

But it will take me a few weeks to learn the v5 (and of course, typescrypt), and for now i don't have free time. So I will have a mixed roadmap, first in Adonis v4 and then migrate to v5 after a couple of months. 

The advantage of this is I will have a real good project for migration and to compare differences, this will help a lot in learning.
